cmd/kubeadm/app/util/etcd/etcd.go

```diff
@@ -176,6 +177,13 @@ func (c Client) AddMember(name string, peerAddrs string) ([]Member, error) {
 		}
 	}
 
+	// Add the new member client address to the list of endpoints
+	parsedPeerAddrs, err := url.Parse(peerAddrs)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error parsing peer address %s", peerAddrs)
+	}
+	c.Endpoints = append(c.Endpoints, GetClientURLByIP(parsedPeerAddrs.Hostname()))
```

Comment:
A few things here:
1. You are parsing the peer address, while the endpoints point to the client address. In most cases they differ only in port number, but let's just be safe here...
2. The call to `url.Parse` (in theory) can fail after successful call to `MemberAdd`, thus failing the `AddMember` while we actually have a new member.
3. You don't need to call `url.Parse` and `GetClientURLByIP`. The client address for the new node is in `resp.Member.ClientURLs`.
4. All of the active members (as of the call to `MemberAdd`) should be returned in `resp.Members`. You can iterate them and get their `ClientURLs` to regenerate the latest version of `c.Endpoints`.

Reply:
Let me explain the process I went through, and why I think this is the less bad solution:

1. on the local etcd setup we [control the endpoints](https://github.com/kubernetes/kubernetes/blob/af31a202ea5ac80ebaca81fa949bf3b31c649618/cmd/kubeadm/app/phases/etcd/local.go#L157) and it's basically going to be this hostname with this port.

2. I didn't think of this case, is this feasible? If the provided peer address is not an URL, how can the member addition succeed?

3. This was my very first try, but the ClientURLs is empty until the `etcd` member starts, and at this point in the code the kubelet is basically starting it. On the response we only have the peer address with an empty list of client URLS, so this is not really an option :(

4. This is the same issue, the new added member is not active and hasn't yet got ClientURLs; it could have (depending if etcd actually started on the background or not).

So I went through the process of checking what options we did have and I tried:

a) Call to `CheckLocalEtcdStatus` directly: the straightforward one. Since this tries to retrieve the endpoints from `GetClusterStatus` on `NewFromCluster` is again a suboptimal situation, because we are blocking trying to download the `kubeadm-config` configmap during the blackout.

b) Call to `Sync()` on etcdclient: during the blackout this again not the best idea, because we would have to loop until we get a valid answer, felt awkward.

c) Try to get the real response from `MemberAdd`: empty clientURLs because the instance could have not started yet; only the peerURLs is populated for the new member.

d) Add the endpoint manually, parsing the hostname and changing the port to the client port.

I also have thought if it's worth checking if this endpoint is on the list to avoid adding duplicates (it wouldn't be a problem, we would check the new endpoint more than once if it was already there, but I was wondering if the hassle of checking if the endpoint exists to avoid checking it twice is worth it).

Reply:
Thanks for the elaboration @ereslibre ! Since you have verified that this is the best working solution, then I am OK to keep it that way.

Reply:
Thank you @rosti for the feedback!